File: LeetcodeQuestions/partition_list-86.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @param {Integer} x
# @return {ListNode}
def partition(head, x)
    return nil if (head == nil)
    
    before_head = ListNode.new(0)
    before_pointer = before_head
    after_head = ListNode.new(0)
    after_pointer = after_head
    
    while head != nil
        if head.val < x
            before_pointer.next = head
            before_pointer = before_pointer.next
        else
            after_pointer.next = head
            after_pointer = after_pointer.next
        end
        head = head.next
    end
    
    after_pointer.next = nil
    before_pointer.next = after_head.next
    return before_head.next
end

File: LeetcodeQuestions/partitionList-86.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} x
 * @return {ListNode}
 */
var partition = function(head, x) {
    
    if (!head) return null;
    
    let beforeHead = new ListNode(0);
    let beforePointer = beforeHead;
    let afterHead = new ListNode(0);
    let afterPointer = afterHead;
    
    while (head) {
        if (head.val < x) {
            beforePointer.next = head;
            beforePointer = beforePointer.next;
        } else {
            afterPointer.next = head;
            afterPointer = afterPointer.next;
        }
        head = head.next;
    }
    
    afterPointer.next = null;
    beforePointer.next = afterHead.next;
    return beforeHead.next;
};